### ISSUE TITLE
CMake build prints "Static linking is broken!"

### ISSUE THREAD

**Issue by zethon** [curl/curl#841]
I want to statically link libcurl into an app I'm writing. I'm using the CMake build system since I'm trying to build with clang. However, the CMake build system is very clear:

> CMake Warning at CMakeLists.txt:1137 (message):
>   Static linking is broken!

However, when I generate and then build I see it's built a file ./lib/libcurl.a

Can this file be trusted?

**Comment by bagder** [comment]
That message was introduced in commit 8ed66f98a95 by @Lekensteyn . You remember the reason? Is it still valid?

**Comment by Lekensteyn** [comment]
This likely refers to the generated curl-config file that is unusable in a certain configuration involving static libs, the actual library could still be functional.

What if you `make install` the library and then try to link with `curl-config --static-libs`? Does that work or will you get an error?

**Comment by zethon** [comment]
`curl-config --static-libs` does work with valid-looking output and no errors.

**Comment by Lekensteyn** [comment]
It seems that the problem is related to the inclusion of other shared libraries:
```
$ curl --libcurl example.c example.com -s -o /dev/null
$ /tmp/curl-root/bin/curl-config --static-libs
/tmp/curl-root/lib/libcurl.a -lc -lidn -llber -lldap -ldl -l/usr/lib64/libssl.so -l/usr/lib64/libcrypto.so -l/usr/lib64/libz.so -l/usr/lib64/libssh2.so
$ gcc example.c `/tmp/curl-root/bin/curl-config --static-libs`
/usr/bin/ld: cannot find -l/usr/lib64/libssl.so
/usr/bin/ld: cannot find -l/usr/lib64/libcrypto.so
/usr/bin/ld: cannot find -l/usr/lib64/libz.so
/usr/bin/ld: cannot find -l/usr/lib64/libssh2.so
collect2: error: ld returned 1 exit status
```

The cmake scripts have to be modified to return a static library (libssl.a, etc.) instead of the shared libs.

**Comment by bagder** [comment]
Static linking of libcurl is however the ability to *build* a static libcurl. *Using* a static libcurl is harder and we actually can't presume that we can figure out and provide the dependency chain of possible libs.

**Comment by zethon** [comment]
Sorry to be using this as a message board for trouble-shooting but you guys seem responsive! :)

I've tried building a simple C++ app with the static library and I am getting several ldap linking errors:

> Undefined symbols for architecture x86_64:
>   "_ldap_url_parse", referenced from:
>       _Curl_ldap in libcurl.a(ldap.c.o)
>   "_ldap_set_option", referenced from:
>       _Curl_ldap in libcurl.a(ldap.c.o)
>   (and on and on)

The thing that is confusing me is that I have disabled LDAP support:

![image](https://cloud.githubusercontent.com/assets/5203583/15796832/53653dd8-29d4-11e6-8a98-e9f64a6f5617.png)

So, I'm not sure why libcurl.a still references LDAP or how to fix this (without having to needlessly link to the LDAP's libs).

Suggestions? Thanks!

**Comment by zethon** [comment]
I've made a little more progress. I was able get past the Linker errors (I think I may have just needed to clear my CMake cache and start from scratch) but now I don't seem to have any HTTP support (I'm on OSX 10.11.3). A simple call to an https site returns `Unsupported protocol`. 

I'm certainly linking to the OpenSSL (and ZLib) libraries when I build libcurl, as well as in my app. 

![image](https://cloud.githubusercontent.com/assets/5203583/15797411/0f7fdd88-29e5-11e6-9875-6646646d9dfc.png)

Thoughts?
